### DISCUSSION TITLE
Use Timers as stopwatch, for counting time intervals

### DISCUSSION THREAD

**Discussion by pgorod** [arendst/Tasmota#13247] [Q&A] [answered]
Hi, I'd like to ask for a bit of help, please.

I have the following rules working well on a Shelly 1PM with Tasmota 9.4.0:

```
Rule 
   ON switch1#state=0 DO websend [192.168.1.250:201] events.php/cmnd=BurnerOff ENDON 
   ON switch1#state=1 DO websend [192.168.1.250:201]  events.php/cmnd=BurnerOn ENDON 
```
This is receiving a signal from a light sensor into the Shelly's SW port and instructing my web server about the on/off events.

My question is **how to keep track of the elapsed time between events**? I know I can do this math on the web server, but it would be useful to me to have that information in the device. 

In the Docs I see Timers, RuleTimers etc., but I couldn't find a way to use these as a stopwatch. I would need to start the stopwatch counting when the switch goes to ON and then send the elapsed time in the WebSend when it goes to OFF.

Is this possible? How?

Thanks!

**Comment by barbudor** [comment]
In rules, you have pseudo variables that can provide informations such as `%time%` which is the number of minutes since midnight : https://tasmota.github.io/docs/Rules/#rule-variables

So you can calculate the elapsed time by rules such as:
```
Rule 
   ON switch1#state=0 DO backlog var1 %time%; websend [192.168.1.250:201] events.php/cmnd=BurnerOff ENDON 
   ON switch1#state=1 DO backlog var2 %time%; websend [192.168.1.250:201]  events.php/cmnd=BurnerOn ENDON 
   ON var2#state DO var3 = var2-var1 ENDON
   ON var3#state DO something with %var3% ENDON
```
You will require a binary including [expressions](https://tasmota.github.io/docs/Rules/#expressions-in-rules)

**Reply by pgorod** [reply]
I won't be using expressions, I need to stick to the current build. Thanks anyway for your detailed answer, much appreciated 🎉 

Still, learning about `%time%` is useful, I might still use this, although I am exploring sfromis' answer now, since I would prefer getting information down to the seconds.

**Reply by pgorod** [reply]
For minute-level measurements only, I was able to get this working:

```
Rule 
   ON switch1#state=0 DO backlog var11 %time%; var12 %time%; sub12 %var11%; event sendweb=%var11% ENDON 
   ON switch1#state=1 DO backlog var11 %time%; var12 %time%; sub12 %var11%; event sendweb=%var11% ENDON 
   ON event#sendweb DO websend [192.168.1.250:201] events.php?cmnd=BurnerOn&now=%time%&previous=%value%&diff=%var12%                 
   ON event#testoff DO backlog var11 %time%; var12 %time%; sub12 %var11%; event sendweb=%var11% ENDON 
   ON event#teston DO backlog var11 %time%; var12 %time%; sub12 %var11%; event sendweb=%var11% ENDON 
ENDON 
```
The two final rules just repeat the previous code, and are meant for tests only. If I knew how to trigger the other events manually, I wouldn't need them.

This works, but it was hard to get right because there are some funky things going on with variables here, I am not sure if it's a bug or just the-way-it-is-due-to-the-unavoidable-limitations-of-Tasmota... The variables only seem to get their values after the whole event handler finishes. So I change %var11%, but later on in the same handler I still get the old value when I reference it...

**Reply by sfromis** [reply]
Yes, the % notation in rules is variables expanded before the do...endon execution starts. Within that scope, it cannot work to change a variable and then expect to use the changed result. This allows things like having a variable contain a command name and/or longer segments of characters, not just simple values. 

If you stick with this approach, better use %uptime% instead of %time% to avoid "midnight trouble".

**Comment by sfromis** [comment, marked as answer]
One way to implement a stopwatch function is to start a ruletimer, with a longer period than the maximum time. At the terminating event, you can then query the remaining time using a RuleTimer command and a JSON trigger, and subtract from the starting period. 

RuleTimer gets you resolution in seconds, while %time% only gives you minutes with the "midnight problem". You can also use %uptime% to have it working cross midnight, still only whole minutes. If you want 1/10 second resolution, you can also do tricks with a fake relay and PulseTime.

Subtractions can also be done without using expressions, using the Sub command, but it is a bit more cumbersome.

**Reply by pgorod** [reply]
Can you please elaborate on this part?

> you can then query the remaining time using a RuleTimer command and a JSON trigger

I need seconds, so I am trying for the RuleTimer solution. I am able to set it running, but I can't figure out the way to read it (and get that on to the WebSend URL).

I think the Sub part I can handle, at least some preliminary tests are giving me what I need.

[I don't know if this is relevant, but I am not using MQTT]

**Reply by sfromis** [reply]
What I wrote was to issue the command RuleTimer from a rule, to make it work in this case as `Backlog RuleTimer`. This will produce a JSON output in MQTT which will also appear in the console. In general, you can pick up values from the published MQTT using a JSON-based trigger like `on T1 do ... endon`. This depends on what levels of objects there are in the JSON.

**Reply by pgorod** [reply]
I'm sorry for being a noob, but after investing a couple of hours on this I still can't get it working.

```
rule3 
   on event#starttimer do backlog sub1 %var2%; RuleTimer8 1000 endon 
   on event#readtimer do backlog RuleTimer endon 
   on RuleTimer#T8 do var5 %value% endon
```
The first two rules work, the third doesn't trigger:
```
18:09:45.597 CMD: event readtimer
18:09:45.605 RSL: RESULT = {"Event":"Done"}
18:09:45.650 RUL: EVENT#READTIMER performs "backlog RuleTimer"
18:09:45.661 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":792}
```
Can you please give me an exact example of how to trigger from a JSON result? Thank you for your patience.

**Reply by sfromis** [reply]
I already gave you an exact example, which you then changed to something very different.

**Reply by pgorod** [reply]
So sorry, you're right, it works exactly as you said; I was being an idiot. This part of the documentation was throwing me off:

![image](https://user-images.githubusercontent.com/15945027/135512594-2ac4deca-b9c0-45d4-90be-bb5fce6b7de2.png)

... because I kept looking for a trigger name, but I see now it's not needed, the valueName of the first level _is_ the triggerName. Phew... 😅 

Working code:
```
rule3 
   on event#starttimer do backlog sub1 %var2%; RuleTimer8 1000 endon 
   on event#readtimer do backlog RuleTimer endon 
   on T8 do var5 %value% endon
```
Now I'll work to get the subtraction working. Thanks a lot and sorry for not understanding your guidance!
